bin/ch-run.c

```diff
@@ -212,40 +274,38 @@ void fix_environment(struct args *args)
    }
 
    // --set-env and --unset-env.
-   for (int i = 0; args->env_deltas[i].action != END; i++) {
+   for (int i = 0; args->env_deltas[i].action != ENV_END; i++) {
       char *arg = args->env_deltas[i].arg;
-      if (args->env_deltas[i].action == SET_FILE) {
-         FILE *fp;
-         Tf (fp = fopen(arg, "r"), "--set-env: can't open: %s", arg);
-         for (int j = 1; true; j++) {
-            char *line = NULL;
-            size_t len = 0;
-            errno = 0;
-            if (-1 == getline(&line, &len, fp)) {
-               if (errno == 0)  // EOF
-                  break;
-               else             // error
-                  Tf (0, "--set-env: error reading: %s", arg);
+      if (args->env_deltas[i].action == ENV_SET) {  // --set-env
+         if (strchr(arg, '=') != NULL) {
+            // argument is variable name & value
+            envs_set((char *[]){ arg }, 1, NULL, args->c.env_expand);
+         } else {
+            // argument is filename
+            char **lines = NULL;
+            int line_ct = 0;
+            FILE *fp;
+            Tf (fp = fopen(arg, "r"), "--set-env: can't open: %s", arg);
+            for (line_ct = 0; true; line_ct++) {
+               char *line;
+               size_t line_len = 0;  // don't care but getline(3) must write
+               errno = 0;
+               if (-1 == getline(&line, &line_len, fp)) {
+                  if (errno == 0)  // EOF
+                     break;        // note: line_ct not incremented
```

Comment:
I couldn't remember whether the third component of the for loop was executed after `break` (here `line_ct++`), and it was surprisingly hard to look up. I had to write a test program. Hence the comment.